[GStreamer] Rework playback rates handling

calvaris opened the pull request:
#### 86301d12f31fa6659af3ee6fc5c0ac8be73b73a7
<pre>
[GStreamer] Rework playback rates handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=259732">https://bugs.webkit.org/show_bug.cgi?id=259732</a>

Reviewed by Alicia Boya Garcia.

There were some use cases in MSE where playback rate change was not working properly and it was decided that a better
idea would be to do a rework of the playback rate handling in the base class. The idea was to not do playback rate state
changes out of the updateStates functions (both regular and MSE) because calling that function should do what is
expected at any moment. In order to do that, we needed to switch the boolean variable representing if pipeline was
paused because of rate into something more reflecting the different states we could be in. We created a self explanatory
enum class (that also was comments to make it even more clear) and set that state when needed for updateStates to move
the pipeline if needed.

A fly-by fix was to make the paused function to return true when we are also moving to paused during an async state
change because the test we are introducing here arose this flaky bug more clearly. The bug was that the media element
went bananas could not realize the player was paused (or on its way to pause) and it was skipping state changes and
playback rate changes.

* LayoutTests/TestExpectations:
* LayoutTests/media/media-source/media-source-play-zero-playbackrate-expected.txt: Added.
* LayoutTests/media/media-source/media-source-play-zero-playbackrate.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::play):
(WebCore::MediaPlayerPrivateGStreamer::pause):
(WebCore::MediaPlayerPrivateGStreamer::paused const):
(WebCore::MediaPlayerPrivateGStreamer::updatePlaybackRate):
(WebCore::MediaPlayerPrivateGStreamer::setRate):
(WebCore::MediaPlayerPrivateGStreamer::updateStates):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::play):
(WebCore::MediaPlayerPrivateGStreamerMSE::pause):
(WebCore::MediaPlayerPrivateGStreamerMSE::updateStates):

Canonical link: <a href="https://commits.webkit.org/268518@main">https://commits.webkit.org/268518@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2395fc61ea58123ca662dfceb82d1da6b2f93e9e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19934 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20981 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21830 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18618 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23616 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20511 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20156 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20127 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17346 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22684 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17297 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18141 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18374 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18317 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/22397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18075 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17977 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4767 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22424 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -292,8 +298,19 @@ void MediaPlayerPrivateGStreamerMSE::sourceSetup(GstElement* sourceElement)
+    if (getStateResult == GST_STATE_CHANGE_ASYNC) {
+        // Changing pipeline state during async state change will be either rejected by changePipelineState()
+        // or ignored by playbin (overwritten by the async tartet state after transitions completes).
+        // Let's wait until it's done and then apply play/pause if needed.
```
This if block makes me a bit uneasy.

> Changing pipeline state during async state change will be either rejected by changePipelineState()

Why? Is that behavior correct?

> or ignored by playbin (overwritten by the async tartet state after transitions completes).

Same questions (why and whether that behavior is correct, a bug of GStreamer or has some sort of rationale I'm missing).

calvaris replied:
I don't really know, never investigated the issue myself.

ntrrgc replied:
Could you test what happens if this if is removed?

calvaris replied:
Nothing happens if it's removed, at least upstream.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -345,6 +345,7 @@ class MediaPlayerPrivateGStreamer : public MediaPlayerPrivateInterface
     mutable std::optional<bool> m_isLiveStream;
     bool m_isPaused { true };
     float m_playbackRate { 1 };
+    bool m_isPlaybackRatePaused { false };
```
This field could use some documentation comment explaining what it represents.

calvaris replied:
It was just moved from below but it is not a problem to add a small comment.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -303,6 +320,11 @@ void MediaPlayerPrivateGStreamerMSE::updateStates()
+    if (getStateResult == GST_STATE_CHANGE_SUCCESS && state >= GST_STATE_PAUSED) {
+        // Update playback rate that was set before the pipeline was prerolled.
+        updatePlaybackRate();
+    }
```
I don't uniderstand this if. `getStateResult` and `state` hold the pipeline state previous to the change. PAUSED->PLAYING is asynchronous, PLAYING->PAUSED is not. That gives us the following possibilities:

* (1) When the function got called the state had been set to PAUSED (`getStateResult == GST_STATE_CHANGE_SUCCESS && state == GST_STATE_PAUSED`).
* <s>(2)</s> When the function got called the state had been set to PLAYING but the pipeline had not pre-rolled yet (`getStateResult == GST_STATE_CHANGE_ASYNC`). However, this case has been eliminated by a previous early return.
* (3) When the function got called the state had been set to PLAYING and the pipeline had already prerolled (`getStateResult == GST_STATE_CHANGE_SUCCESS && state == GST_STATE_PLAYING`).
* <s>(4)</s> Playback has ended with a state change to READY or NULL. I don't know of a case where this happens in the MSE player.

So this seems like a condition that would always be executed. I'm also confused because it checks the old state rather than the new but doesn't explain why, I don't get it.

calvaris replied:
There are other state changes than success and async. I guess it makes sense to not do anything in any other than success. Then, if we have already a state change sinked, i.e. success, what we want is to call `updatePlaybackRate` so that it can move the pipeline to playing if the rate is suitable for that.

ntrrgc replied:
> There are other state changes than success and async.

Although they exist in the abstract, I don't see how they could happen in this particular case, and if they can, it should be explained.

ntrrgc commented on LayoutTests/media/media-source/media-source-play-zero-playbackrate.html:
```diff
@@ -0,0 +1,92 @@
+<html>
+<head>
+    <title>media-source-play-zero-playbackrate</title>
+    <script src="media-source-loader.js"></script>
```
Add this line:

```
<meta name="timeout" content="long">
```

Since this test has intentional waits in the order of seconds, it's a good idea to mark it as a long test. This is not a big deal when running the test alone in a relatively idle machine, but it leaves the test timeout with very little buffer to cover any hiccup in the bots from running many more processes.

ntrrgc reviewed (commented):
Given we're both pretty busy and that this patch fixes a bug and has a test, I'm open to accepting a revised version that preserves some of the unknowns mentioned, as long as they're documented as such and the patch has been tested to not cause any regressions. Of course, I would very much prefer those unknowns to be resolved and explained in code comments.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -303,6 +321,12 @@ void MediaPlayerPrivateGStreamerMSE::updateStates()
+        // Update playback rate that was set before the pipeline was prerolled.
+        // FIXME: This needs further investigation in https://bugs.webkit.org/show_bug.cgi?id=259776.
+        updatePlaybackRate();
+    }
```
I've dig a bit further, I found this exact condition in `MediaPlayerPrivateGStreamer::updateStates()`, so I would adventure it may have been copied verbatim:

    if (getStateResult == GST_STATE_CHANGE_SUCCESS && m_currentState >= GST_STATE_PAUSED) {
        updatePlaybackRate();

Now, what does `updatePlaybackRate()` do?

1. Returns early if `!m_isChangingRate`.
2. Logs about the rate change about to occur.
3. Decides whether to mute audio depending on the rate set.
4. If `m_lastPlaybackRate != m_playbackRate`, calls `doSeek(playbackPosition(), m_playbackRate)`. Both regular and MSE implementations of `doSeek()` would rewrite a playback rate of 0 to 1 here, were it to occur. I suspect the goal here is to not do a seek if pausing/unpausing with rates != 0?
5. If `m_isPlaybackRatePaused` is set, it will be consumed (set to false) and the pipeline will be set to PLAYING if it wasn't already. **Therefore, there is therefore an implicit assumption that `updatePlaybackRate()` must not be called when `m_playbackRate == 0 && m_isPlaybackRatePaused`.** Indeed `MediaPlayerPrivateGStreamer::setRate()` will return early on rate == 0 to avoid that, just after setting the pipeline to PAUSED.
6. `m_isChangingRate` is reset to FALSE and `m_player->rateChanged()` is called, which will update the `m_reportedPlaybackRate` of the HTMLMediaElement with whatever `effectiveRate()` returns, which defaults to whatever `rate()` returns.

All this rate handling code looks quite messy to me. It's very hard for me to analyze. State changes are happening from many different places.

One goal with my MediaPlayerPrivateGStreamerMSE seek rework was cleaning up and simplifying it, which can be clearly seen by comparing `MediaPlayerPrivateGStreamerMSE::updateStates()` to `MediaPlayerPrivateGStreamer::updateStates()`.

I think the rate handling code could use a cleanup and/or refactor.

calvaris replied:
After your analysis, surely you will correct me if I am wrong, I think the best case is keeping this but ensuring we are going going though that if playback is stopped because of zero rate. I updated the PR with that.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -140,13 +140,19 @@ void MediaPlayerPrivateGStreamerMSE::play()
+    if (!m_playbackRate) {
+        // If the playback rate is 0, we should not resume playback.
+        m_isPlaybackRatePaused = true;
+        return;
```
`updateStates()` should recognize `m_isPlaybackRatePaused` is set and not change the pipeline state, so why the return?

The whole point of the `updateStates()` in MediaPlayerPrivateGStreamerMSE is that it will use all information to determine if the pipeline state needs to be changed. By returning early, the `play()` function is distrusting `updateStates()` and breaking consistency in this particular case.

calvaris replied:
No, it is not recognizing that. Hence, this is the point where we set this variable to true, which keeps track of the condition that a play was issued while rate == 0.

ntrrgc replied:
I was referring specifically to the `return`, not to `m_isPlaybackRatePaused` (which is also a suspicious variable, tbf, but beside that point).

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -303,6 +313,11 @@ void MediaPlayerPrivateGStreamerMSE::updateStates()
+
+    if (!m_isPlaybackRatePaused && getStateResult == GST_STATE_CHANGE_SUCCESS && state >= GST_STATE_PAUSED) {
+        // Update playback rate that was set before the pipeline was prerolled.
+        updatePlaybackRate();
```
What does `updatePlaybackRate()` and in what situations do we need to call it?

calvaris replied:
`updatePlaybackRate()` accounts for `m_playbackRatePaused` and moves the pipeline to playing if needed, among other things. That's why we need to call it in this situation.

ntrrgc replied:
But isn't that already the responsibility of `updateStates()`, just in the previous if() statement?

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -345,6 +345,7 @@ class MediaPlayerPrivateGStreamer : public MediaPlayerPrivateInterface
     mutable std::optional<bool> m_isLiveStream;
     bool m_isPaused { true };
     float m_playbackRate { 1 };
+    bool m_isPlaybackRatePaused { false }; // Marks if the player is paused because the playback rate was set to 0. We don't move to playing in this case.
```
What should be the value of m_isPlaybackRatePaused if we call .pause() while playbackRate == 0? This comment doesn't make it clear.

ntrrgc reviewed (changes requested):
After my last examination, I conclude the rate changing code needs further changes so that  `MediaPlayerPrivateGStreamerMSE::updateStates()` and the playback rate handling code don't compete for the responsibility of managing the pipeline state.

This will likely require some level of refactor or rework in the parent class playback rate handling code. Additional cleanups are appreciated.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -433,15 +434,17 @@ bool MediaPlayerPrivateGStreamer::paused() const
 
-    if (m_isPlaybackRatePaused) {
+    if (m_playbackRatePausedState == PlaybackRatePausedState::RatePaused
+        || m_playbackRatePausedState == PlaybackRatePausedState::ShouldMoveToPlaying) {
```
When can we be in PlaybackRatePausedState::ShouldMoveToPlaying? So far that one seems like a virtual state the player does never stay on.

calvaris replied:
It states that a transition to playing is needed as we don't do pipeline state changes out of updateStates anymore.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -387,7 +387,8 @@ void MediaPlayerPrivateGStreamer::play()
 
     if (!m_playbackRate) {
-        m_isPlaybackRatePaused = true;
+        if (m_playbackRatePausedState == PlaybackRatePausedState::ManuallyPaused)
```
Why is this conditional?

calvaris replied:
To prevent changes from ShouldMoveToPlaying in case we do:
* rate = 0
* rate = 1
* play()

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -433,15 +434,17 @@ bool MediaPlayerPrivateGStreamer::paused() const
-    GST_LOG_OBJECT(pipeline(), "Paused: %s", toString(paused).utf8().data());
+    GstState state, pending;
+    auto stateChange = gst_element_get_state(m_pipeline.get(), &state, &pending, 0);
+    bool paused = state <= GST_STATE_PAUSED || (stateChange == GST_STATE_CHANGE_ASYNC && pending == GST_STATE_PAUSED);
```
The change in this boolean expression makes sense and I can see it fixing a bug. Is this a drive-by fix?

ntrrgc replied:
Yes it is, it is stated in the commit message.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -255,6 +255,13 @@ class MediaPlayerPrivateGStreamer : public MediaPlayerPrivateInterface
 
+    enum class PlaybackRatePausedState {
+        ManuallyPaused, // Initialization or user explicitly paused.
+        RatePaused, // Pipeline was playing and rate was set to zero.
```
Which state has priority? If we're both paused with `.pause()` and with `.rate = 0`, are we `ManuallyPaused` or `RatePaused`? Do we get the same state if we do `.pause()` first and `.rate = 0` later than if we do `.rate = 0` first and `.pause()` second?

ntrrgc replied:
calvaris: `ManuallyPaused` has priority <-- should be in the comment

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -255,6 +255,13 @@ class MediaPlayerPrivateGStreamer : public MediaPlayerPrivateInterface
+    enum class PlaybackRatePausedState {
+        ManuallyPaused, // Initialization or user explicitly paused.
+        RatePaused, // Pipeline was playing and rate was set to zero.
+        ShouldMoveToPlaying, // Pipeline was paused because of zero rate and it should be playing.
```
This looks a lot like a virtual state: one that, given a call from JavaScript we will never be when we return to user code, but that rather is used as a sneaky way to modulate the behavior of functions that you didn't want to change further (e.g. by adding explicit arguments to them).

ntrrgc replied:
^ This detail should be documented in the comment.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -140,13 +140,16 @@ void MediaPlayerPrivateGStreamerMSE::play()
 {
     GST_DEBUG_OBJECT(pipeline(), "Play requested");
     m_isPaused = false;
+    if (!m_playbackRate && m_playbackRatePausedState == PlaybackRatePausedState::ManuallyPaused)
```
Given the case of `m_playbackRate == 0` when `play()` is called, can we be in `PlaybackRatePausedState::Playing`? If so, how? If not, as I would assume, that would make for a good assert.

What other states can we be in here? If they are `ManuallyPaused` and `RatePaused`, both would do the same outcome, making the second part of the `if` redundant.

calvaris replied:
In the first case, it can't happen. I guess you mean we should assert before the if.

The other states can happen but not all of them with `m_playbackRate == 0`.

ntrrgc reviewed (approved):
I still think the rate and pause handling could be further simplified. Anyway, r+ with some documentation requests.

calvaris committed:
[GStreamer] Rework playback rates handling
https://bugs.webkit.org/show_bug.cgi?id=259732

Reviewed by Alicia Boya Garcia.

There were some use cases in MSE where playback rate change was not working properly and it was decided that a better
idea would be to do a rework of the playback rate handling in the base class. The idea was to not do playback rate state
changes out of the updateStates functions (both regular and MSE) because calling that function should do what is
expected at any moment. In order to do that, we needed to switch the boolean variable representing if pipeline was
paused because of rate into something more reflecting the different states we could be in. We created a self explanatory
enum class (that also was comments to make it even more clear) and set that state when needed for updateStates to move
the pipeline if needed.

A fly-by fix was to make the paused function to return true when we are also moving to paused during an async state
change because the test we are introducing here arose this flaky bug more clearly. The bug was that the media element
went bananas could not realize the player was paused (or on its way to pause) and it was skipping state changes and
playback rate changes.

* LayoutTests/TestExpectations:
* LayoutTests/media/media-source/media-source-play-zero-playbackrate-expected.txt: Added.
* LayoutTests/media/media-source/media-source-play-zero-playbackrate.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::play):
(WebCore::MediaPlayerPrivateGStreamer::pause):
(WebCore::MediaPlayerPrivateGStreamer::paused const):
(WebCore::MediaPlayerPrivateGStreamer::updatePlaybackRate):
(WebCore::MediaPlayerPrivateGStreamer::setRate):
(WebCore::MediaPlayerPrivateGStreamer::updateStates):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::play):
(WebCore::MediaPlayerPrivateGStreamerMSE::pause):
(WebCore::MediaPlayerPrivateGStreamerMSE::updateStates):

Canonical link: https://commits.webkit.org/268518@main
